Blog Post: NumPy Benchmarking (#260)
Quansight-Labs/quansight-labs-site: posts/2021/10/numpy_benchmarking_blog.md

- **Compiled Methods**: This part was the most exciting part of the project. 
I got introduced to various compiled methods like [Cython](https://cython.readthedocs.io/en/latest/), 
[Pythran](https://pythran.readthedocs.io/en/latest/), [Numba](http://numba.pydata.org/), [Transonic](https://transonic.readthedocs.io/en/latest/). 
It was the first time when I got to know about accelerators. 
I visited their documentations, GitHub, and searched popular blogs about compiled methods. 
I loved playing with them. I read the theory part, looked into the examples, 
and implemented them in my editor. It was a great learning experience for me 
to get familiar with compiled methods. 
I implemented the jitted compilation mode in Pythran and Numba for benchmarking via Transonic's support.

- **Visualization**: I used the [Matplotlib](https://matplotlib.org/) library for visualizing the benchmarking results. 
I tried various plots to verify which one suits best, 
like scatter plots, box plots, line charts, the combination of the scatter plots and line charts, etc. 
But these were not good to go. 
Those plots either lacked clarity or were not capable of providing significant results. 

> We finalized decided on horizontally stacked bar charts.

Reviewer: I thought we are using a [vertical grouped bar chart](https://matplotlib.org/stable/gallery/index.html)
```suggestion
We finalized decided on two bar charts, with different vertical scales to accomodate the vastly different performance of Python vs. the compiled methods. We also normalized the data to show trends as the number of particles increased.
```

- **Model Optimization**: Model Optimization was one of the most exciting parts for me to work. 
I like playing with codes. The main task was to ensure 
we were obtaining similar results in all the implemented algorithms. 
I revisited all the code I had executed earlier. At this stage, 
I was able to find out errors in my code and had an idea to improve it. 
The final aim was to achieve the same results in each step at a minimal time. 
Steps I followed to attain it:
	
	- Initially, I played around with the library functions to check out which library function gave the best results.
	
	- I then turned my focus to reduce the number of loops. 
And I'll say hats off to the *Vectorized Approach* of the NumPy. 

> NumPy achieved more than 10% than Python and 

Reviewer: 10% of what?

Author: Ah, yes! Thanks for pointing it out.

I corrected that to:
> NumPy achieved a speed of more than 10% faster than Python. The compelling thing is the changing behavior of NumPy from python-like performance to compiled-like performance.

equivalent speed to compiled methods like C++ and Numba.
	
	- The only task left was to verify whether we got the same results in all the cases. 

> Initially, I wanted to make my code as small as possible. 

Reviewer: But was it really faster? The difference in speed between `np.add(a, b)` and `a + b` was negligible, if measurable at all.
```suggestion
Initially, I wanted to make my code as fast as possible. 
```

Author: > But was it really faster? The difference in speed between np.add(a, b) and a + b was negligible, if measurable at all.

Yes, you are correct those differences were negligible and were varying. 

Here I was talking about the non-trivial approach I used earlier.
This one: https://github.com/numpy/numpy.org/pull/461#discussion_r705694228

Does that make sense?

<!--
.. title: NumPy Benchmarking
.. slug: numpy-benchmarking
.. date: 2021-09-27 12:23:40 UTC-05:00
.. author: Khushi Agrawal
.. tags: NumPy, Pythran, Numba, Transonic
.. category: 
.. link: 
.. description: 
.. type: text
.. previewimage:
-->

In this blog post, I'll be talking about my journey in Quansight. I want to share all things I was involved in and accomplished. What issues I faced, and most importantly, what were awesome life hacks I learned during this period.

First of all, I'd like to express my gratitude to the whole team for allowing me to be a part of such a great team. My work was majorly focused on providing performance benchmarks to NumPy in realistic situations. The target was to show the world that NumPy is efficient in handling quasi real-life situations too.

<p align="center">
      <img src = "../../images/2021/10/journey.jpeg" alt = "My Journey!">

> </p>

Reviewer 2: I think this is a good suggestion! Reading through the blog post, it seems like all of these things are mentioned so the content isn't necessarily missing. If I do think about this in terms of reading order, though, I think it might better preserve the kind of preview nature of the image to add the following:

<img src = "/images/2021/10/journey.jpeg" alt = "A word cloud with themes, open-source projects and people mentioned throughout the blog post. Each is stylized using a different font, most of them calligraphical.">

As always, this is just a suggestion! I think what is there now is a clear description, just less linked to the rest of the post.

- **Search Everything about your project in the first 3-4 days**: At this part, you need to get familiar with all the possible dots of your project. Look into as many related works of your project and examine the positive and negative points of the proposed work. Now it's high time to give structure to your project. I was pretty much sure about my work. After getting familiar with the problem statement, I read various other proposed projects related to benchmarking. Fewer of them were [initialcontiditions.org](http://initialconditions.org/), [benchmarks game](https://benchmarksgame-team.pages.debian.net/benchmarksgame/), [Julia's micro-benchmarks](https://julialang.org/benchmarks/); there were a few more. I agree that it took more than three days for me to complete, but I learned specific life hacks, which I'm damn sure that I will implement in every project. Make sure not to dig deeper into the topic, first know the width, then dive into the depth and ensure that you are stumbling upon the subject-related issues.

> - **Start working**: Here come the journey starts. The best way to express yourself is to present Everything that you have completed. Ask doubts as much as you can. But make sure that you have spent quality time in it. I used to update my mentor Matti Picus each day about the progress of our work. I am so glad to get such a very responsive and understanding mentor.

Reviewer 3: I think you mean
```suggestion
- **Start working**: Here is where the journey starts. The best way to express yourself is to present Everything that you have completed. Ask doubts as much as you can. But make sure that you have spent quality time in it. I used to update my mentor Matti Picus each day about the progress of our work. I am so glad to get such a very responsive and understanding mentor.
```

<!--
.. title: NumPy Benchmarking
.. slug: numpy-benchmarking
.. date: 2021-09-27 12:23:40 UTC-05:00
.. author: Khushi Agrawal
.. tags: NumPy, Pythran, Numba, Transonic
.. category: 
.. link: 
.. description: 
.. type: text
.. previewimage:
-->

In this blog post, I'll be talking about my journey in Quansight. I want to share all things I was involved in and accomplished. What issues I faced, and most importantly, what were awesome life hacks I learned during this period.

First of all, I'd like to express my gratitude to the whole team for allowing me to be a part of such a great team. My work was majorly focused on providing performance benchmarks to NumPy in realistic situations. The target was to show the world that NumPy is efficient in handling quasi real-life situations too.

<p align="center">
      <img src = "/images/2021/10/journey.jpeg" alt = "A word cloud with themes, open-source projects and people mentioned throughout the blog post. Each is stylized using a different font, most of them calligraphical.">
</p>

> 

Reviewer 3: You should also include a teaser marker around here for the blog main page.
```suggestion

<!-- TEASER_END -->

```

## My Experience
My work was broadly divided majorly into the following categories:

> - **N-Body Problem**: The [N-Body problem](https://en.wikipedia.org/wiki/N-body_problem) is one of the most famous and universally accepted problems for benchmarking. I was given this as a problem statement to work on the project. I started my work with a theoretical understanding of the problem. It was a fun learning part for me to connect the scientific part with the programming world. I implemented the N-Body problem in Python, C, and C++.

Reviewer: Please give credit for the original code with a link to the repo. Something like "I adopted the code from ..."

Author: Yes, sure!

- **Issue**: The most interesting issue I faced was using the vectorized approach in Pythran's implementation. I mentioned that [here](https://github.com/khushi-411/numpy-benchmarks/issues/4).

## Other Technical Work

- **Benchmarking Environment**: I enjoy changing my OS and love to taste different environments. 
But it was the first time I isolated a certain number of CPU cores for accurate benchmarking results. 
I referred to the official documentation of [pypref](https://pyperf.readthedocs.io/en/latest/) 
and visited more than ten blogs to understand the idea. 
It was a fun learning part.

- **Git**: Getting familiar with various git commands was one of the most incredible things 

> I became comfortable with while working at Quansight.

Reviewer 3: ```suggestion
I became comfortable with it while working at Quansight.
```

## Advice to the Beginners

- **Getting Familiar with the importance of the project**: I believe: 

> 'To find joy in the work the most important task is to know, where it started from.' 

Reviewer 3: ```suggestion
'To find joy in the work, the most important task is to know where it started from.' 
```

Read the previous discussions made and know the reason for the importance of your project. 
I started my work with its origin. I read the issue related to benchmarking, 
articles, and other related work. 

> I visited benchmarking pages of other libraries too to get the idea. 

Reviewer 3: ```suggestion
I visited benchmarking pages of other libraries, too, to get the idea. 
```

Among which the [micro-benchmarks of NumPy](https://pv.github.io/numpy-bench/) using ASV are the best. 
It's too lovely!

> - **Search Everything about your project in the first 3-4 days**: At this part, 

Reviewer 3: ```suggestion
- **Search everything about your project in the first 3-4 days**: At this part, 
```

you need to get familiar with all the possible dots of your project. 
Look into as many related works of your project and examine 
the positive and negative points of the proposed work. 
Now it's high time to give structure to your project. 
I was pretty much sure about my work. 
After getting familiar with the problem statement, 
I read various other proposed projects related to benchmarking. 
A few of them were [initialcontiditions.org](http://initialconditions.org/), 
[benchmarks game](https://benchmarksgame-team.pages.debian.net/benchmarksgame/), [Julia's micro-benchmarks](https://julialang.org/benchmarks/); 
there were a few more. 
I agree that it took more than three days for me to complete, 
but I learned specific life hacks, which I'm pretty sure 
that I will implement in every project. 

> Make sure not to dig too deep into the topic, 
> first know the width, then dive into the depth and 
> ensure that you are stumbling upon the subject-related issues.

Reviewer 3: I have a slight rephrasing suggestion:
```suggestion
Make sure not to dig too deep into the topic. First, know the width, then dive into the depth and 
ensure that you are focusing on the subject.
```
